doc/man3/SSL_CTX_set_tlsext_servername_callback.pod

```diff
@@ -34,8 +34,29 @@ the application callback.
 SSL_CTX_set_tlsext_servername_arg() sets a context-specific argument to be
 passed into the callback for this B<SSL_CTX>.
 
-SSL_get_servername() returns a servername extension value of the specified
-type if provided in the Client Hello or NULL.
+The behaviour of SSL_get_servername() depends on a number of different factors.
+In particular note that in TLSv1.3 the servername is negotiated in every
+handshake. In TLSv1.2 the servername is only negotiated on initial handshakes
+and not on resumption handshakes.
+
+On the client, if this is called before a server accepts or ignores the
+servername, then it will return the servername being requested if one has been
+set. If a servername has not been set then it returns the servername from the
+original session (if a resumption is being attempted). After the server has
+accepted or ignored the servername, if TLSv1.2 or below has been negotiated and
+a session was resumed, then it returns the servername from the original session
+if there was one. Otherwise the servername requested by the client in this
+handshake is returned.
+
+On the server, if this is called before the servername extension from the client
+is processed, then NULL will be returned. After processing of any servername
+extension from the client, if TLSv1.2 has been negotiated and a session was
+resumed, then the servername from the original session is returned. Otherwise
+the servername requested by the client in this handshake is returned.
+
+Note that the ClientHello callback occurs before a servername extension from the
+client is processed. The servername and alpn callbacks occur after a
+servername extension from the client is processed.
```

Comment:
Note that the first thing this manpage says is:
`The functionality provided by the servername callback is superseded by the ClientHello callback`

But now you're saying it can't actually be used.

I think it's also useful to mention when the callback set using SSL_CTX_set_cert_cb() is called, here or in the NOTES section.

Reply:
> Note that the first thing this manpage says is:
The functionality provided by the servername callback is superseded by the ClientHello callback
>
> But now you're saying it can't actually be used.

No, you can't use SSL_get_servername() in the client hello callback. This is by design. The `SSL_get_servername()` function was designed for use with the old servername callback. The client hello callback on the other hand works differently and occurs much earlier in the process - before client hello extensions are processed. The client hello callback documentation explains how you are supposed to access extension data in this case.

> I think it's also useful to mention when the callback set using SSL_CTX_set_cert_cb() is called, here or in the NOTES section.

Added.

Reply:
> The functionality provided by the servername callback is superseded by the ClientHello callback

Hmm....although...this sentence is not entirely true because you actually still need to use the servername callback even if most of the work is done by the client hello callback. You need the servername callback to be present in order for the server to acknowledge the SNI extension.

Reply:
I've updated the documentation to try and capture this and to expand out more on how a servername callback is supposed to work.